### ISSUE TITLE
MathML conversion not happening in general

### ISSUE THREAD

**Issue by opoudjis** [plurimath/plurimath#105]
`Plurimath::Math.parse("sum_(i=1)^n i^3=((n(n+1))/2)^2", "asciimath")`

works fine, but

`Plurimath::Math.parse("sum_(i=1)^n i^3=((n(n+1))/2)^2", "asciimath").to_mathml`

fails:

````
undefined method `to_mathml_without_math_tag' for {:expr=>{:power_base=>#<Plurimath::Math::Function::Sum:0x0000000108cbf580 @parameter_one=#<Plurimath::Math::Formula:0x0000000108a3b3e0 @value=[#<Plurimath::Math::Symbol:0x0000000108e4e3d8 @value="i">, #<Plurimath::Math::Symbol:0x0000000108edfbd0 @value="=">, #<Plurimath::Math::Number:0x0000000108f05308 @value="1">]>, @parameter_two=#<Plurimath::Math::Symbol:0x0000000108cc6d30 @value="n">>, :expr=>{:expr=>{:power=>#<Plurimath::Math::Function::Power:0x0000000108d4cea8 @parameter_one=#<Plurimath::Math::Symbol:0x0000000108d4cf48 @value="i">, @parameter_two=#<Plurimath::Math::Number:0x0000000108d4ced0 @value="3">>, :expr=>{:sequence=>#<Plurimath::Math::Symbol:0x0000000108d6c618 @value="=">, :expr=>{:power=>{:intermediate_exp=>#<Plurimath::Math::Formula:0x0000000108d641c0 @value=[#<Plurimath::Math::Formula:0x0000000108ab2b70 @value=[#<Plurimath::Math::Symbol:0x0000000108dec4d0 @value="n">, #<Plurimath::Math::Formula:0x0000000108eec290 @value=[#<Plurimath::Math::Symbol:0x0000000108e0ca78 @value="n">, #<Plurimath::Math::Symbol:0x0000000108e86c60 @value="+">, #<Plurimath::Math::Number:0x0000000108eac280 @value="1">]>]>, #<Plurimath::Math::Symbol:0x0000000108ca41b8 @value="/">, #<Plurimath::Math::Number:0x0000000108cc4f58 @value="2">]>, :^=>"^", :number=>"2"}}}}}}}:Hash (NoMethodError)

        value.map(&:to_mathml_without_math_tag).join
             ^^^^
	from /Users/nickn/.rbenv/versions/3.1.3/lib/ruby/gems/3.1.0/gems/plurimath-0.2.1/lib/plurimath/math/formula.rb:35:in `mathml_content'
````

**Comment by suleman-uzair** [comment]
@opoudjis https://github.com/plurimath/plurimath/issues/103#issuecomment-1383602953

**Comment by opoudjis** [comment]
Converts:

````
<math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
  <mstyle displaystyle="true">
    <mrow>
      <mo>&#x2211;</mo>
      <mrow>
        <mi>i</mi>
        <mi>&#x3d;</mi>
        <mn>1</mn>
      </mrow>
      <mi>n</mi>
    </mrow>
    <msup>
      <mi>i</mi>
      <mn>3</mn>
    </msup>
    <mi>&#x3d;</mi>
    <msup>
      <mfrac>
        <mrow>
          <mi>n</mi>
          <mrow>
            <mo>(</mo>
            <mi>n</mi>
            <mi>&#x2b;</mi>
            <mn>1</mn>
            <mo>)</mo>
          </mrow>
        </mrow>
        <mn>2</mn>
      </mfrac>
      <mn>2</mn>
    </msup>
  </mstyle>
</math>
````

![image](https://user-images.githubusercontent.com/1911039/213167523-4bb0654d-e8eb-4164-9d67-c4344f29ae25.png)


The output of the asciimath gem is:

````
<math>
	<munderover>
		<mrow>
			<mo>&#x2211;</mo>
		</mrow>
		<mrow>
			<mrow>
				<mi>i</mi>
				<mo>=</mo>
				<mn>1</mn>
			</mrow>
		</mrow>
		<mrow>
			<mi>n</mi>
		</mrow>
	</munderover>
	<msup>
		<mrow>
			<mi>i</mi>
		</mrow>
		<mrow>
			<mn>3</mn>
		</mrow>
	</msup>
	<mo>=</mo>
	<msup>
		<mrow>
			<mfenced open="(" close=")">
				<mrow>
					<mfrac>
						<mrow>
							<mrow>
								<mi>n</mi>
								<mfenced open="(" close=")">
									<mrow>
										<mi>n</mi>
										<mo>+</mo>
										<mn>1</mn>
									</mrow>
								</mfenced>
							</mrow>
						</mrow>
						<mrow>
							<mn>2</mn>
						</mrow>
					</mfrac>
				</mrow>
			</mfenced>
		</mrow>
		<mrow>
			<mn>2</mn>
		</mrow>
	</msup>
</math>
````

![image](https://user-images.githubusercontent.com/1911039/213167722-d8b18c89-9110-43ac-b3e7-8d2d722338b7.png)

So, yeah, we do need work on MathML basics, before we can do testing with the samples. The concern is that, unlike the MathML > AsciiMath direction, the AsciiMath > MathML direction is more fragile semantically, so that expressions may well parse correctly but display incorrectly, with us none the wiser.

**Comment by opoudjis** [comment]
There are only two things missing in the MathML parse, but they are vital:

* That the arguments of the Sum are stacked as munderover
* And the relative ordering of the parentheses and the exponentiation

Ronald suggests incorporating this formula into the rspec testing. (It is the test formula of asciimath.org)

**Comment by ronaldtse** [comment]
I suggest we:

1. Gather a list of formulas in asciimath
2. Create a spec for each of them
3. In the spec, for every formula, obtain the MathML output from plurimath and the output from the asciimath gem.
4. If the outputs are different, fail the spec.

This allow us to properly test interoperability with the asciimath gem.

**Comment by ronaldtse** [comment]
FYI mind you that asciimath.org uses asciimath.js which produces DIFFERENT output from the asciimath gem.

**Comment by opoudjis** [comment]
There are going to be minor differences in rendering of MathML, which map to the same visual presentation. I would not get too worried about those (though it does mean we may end up altering the asciimath gem output for testing, if the differences are indeed minor.)

**Comment by suleman-uzair** [comment]
@ronaldtse 

We implemented **OMML** conversion based on the [**scienceai/omml2mathml**](https://github.com/scienceai/omml2mathml) library which provides around 178 examples of **OMML**, which didn't cover all the scenarios we have in **inter-lingual conversion**.

So adding conversion methods for all functions(function classes) will need some more time to implement.
In **AsciiMath** and **Latex** we ignored parenthesis(as said [**here**](https://github.com/plurimath/plurimath/pull/7#discussion_r834930299)) which is causing [**this issue**](https://github.com/plurimath/plurimath/issues/105)(and i'm working on it), which requires some changes to be done manually in each class(or parent classes if common for most).

And [here](https://github.com/plurimath/plurimath/issues/105#:~:text=In%20the%20spec%2C%20for%20every%20formula%2C%20obtain%20the%20MathML%20output%20from%20plurimath%20and%20the%20output%20from%20the%20asciimath%20gem.), as you mentioned to use the [**asciimath gem**](https://rubygems.org/gems/asciimath) for comparison and @opoudjis raised a good point as well.

So I'm adding specs([an example](https://github.com/plurimath/plurimath/pull/110/files#:~:text=describe%20%22.to_latex%2C%20to_mathml,.to_formula%20%7D)) for each representational language(**MathML**, **AsciiMath**, and **Latex**) based on the comparison of presentation of each library's output(comparing the output of **Plurimath** and [**asciimath gem**](https://rubygems.org/gems/asciimath)).

**Comment by ronaldtse** [comment]
@suleman-uzair this makes sense, when's the scheduled completion of this work?

**Comment by suleman-uzair** [comment]
@ronaldtse 

Some examples are taking a decent time for making all the changes and I also have to compare the output(if I change something for any base class, that needs an update in a lot of files, so this is the reason it’s consuming time and I can’t say anything with certainty)
But I will be trying to push fixes/changes in here on daily basis: https://github.com/plurimath/plurimath/pull/110
Right now, I am trying to target long & complex equations to reduce time consumption and unnecessary overwork.

Should we cover all examples here?

**Comment by ronaldtse** [comment]
I believe we should cover as many reasonable examples as possible. Can you maintain a list of examples that you have not integrated yet so we can track progress?

**Comment by suleman-uzair** [comment]
@ronaldtse 

> - |(a,b),(c,d)|=ad-bc
> - ((a_(11), cdots , a_(1n)),(vdots, ddots, vdots),(a_(m1), cdots , a_(mn)))
> - "Скорость"=("Расстояние")/("Время")
> - bb (a + b) + cc c = fr (d^n)
> - text("foo")
> - ubrace(1 + 2) obrace(3 + 4
> - s\'_i = {(- 1, if s_i > s_(i + 1)),( + 1, if s_i <= s_(i + 1)):}
> - overset (a + b) (c + d)
> - underset a b
> - cancel(a_b^c) cancel a_b^c
> - color(red)(x) color(red)(y) color(blue)(z) colortext(blue)(a_b^c)
> - { x\ : \ x in A ^^ x in B }
> - [[a,b,|,c],[d,e,|,f]]
> - ubrace(((1, 0),(0, 1)))_("Adjustment to texture space")
> - ii(V)(ii(X)) = (b-a)^2/12 + d^2/9.
> - "CO"_2" (aq.) "+ "H"_2"O" overset(larr)(rarr) "HCO"_3^(-) + "H"^(+) " log "ii(K) = -6.38 80_{0}^{+2}
> - u(w_("cal")) = w_("cal") *sqrt([(u(m_("stock")))/m_("stock")]^2 +[(u(b_("stock")))/b_("stock")]^2 +[(u(w_("stock")))/w_("stock")]^2 +[(u(m_("sol")))/m_("sol")]^2 +[(u(b_("sol")))/b_("sol")]^2 +2*[(u(ii(V)))/ii(V)]^2)
> - "1 "unitsml(A)" = ( ((4pi xx 10^(-7)))/((9192631770)(299792458)(1)) )^(1/2) ( (Delta ii(nu)_("Cs")cm_(cc K))/(ii(mu)_0) )^(1/2) = 6.789687... xx 10^(-13) ((Delta ii(nu)_("Cs")c m_(cc K))/(ii(mu)_0))^(1/2)"
> - ""_rZZ_n = sum_(j=1)^n (-1)^j j^r, " with " r = 1,2,...,
> - (ii(T)_1+ii(T)_2-a_1-a_2)*f(t) = {(0,pour,{(0&#x3c;t&#x3c;a_(-)),("ou "t&#x3e;ii(T)_(-)+a_2):}),(1,"''",{(a_(-)&#x3c;t&#x3c;a_+","),(ii(T)_1-a_1&#x3c;t&#x3c;ii(T)_(-)),("ou "ii(T)_1&#x3c;t&#x3c;ii(T)_(-)+a_2):}),(2,"''",a_+&#x3c;t&#x3c;ii(T)_1-a_1),((ii(T)_3-ii(T)_(-)+a_1)*delta(t-ii(T)_1),pour,t=ii(T)_1),((ii(T)_1-ii(T)_4)*delta(t-ii(T)_2),"''",t=ii(T)_2","):}
> - ul (ii(B)) = [[1,,,,], [1,1,,,], [1,2,1,,], [color(red)(1),color(red)(3),color(red)(3),color(red)(1),], [1,4,6,4,1], [,,...,,]]
> - H_s(z)=(1-A(ztext(/)overset(~)(gamma)))/(1-A(ztext(/)overset(~)(gamma)))(1+mu z^(-1)), text( with ) 0&#x3c;overset(~)(gamma)_1&#x3c;overset(~)(gamma)_2&#x3c;=1
> - &lt; &lt; mode, &lt; s(ag) &gt; &gt;, s
> - {:(u^2(delta y), = {:[({partial f}/{partial ii(X)_1})^2 u^2(x_1) +({partial f}/{partial ii(X)_2})^2 u^2(x_2) +2 {partial f}/{partial ii(X)_1} {partial f}/{partial ii(X)_2} r(x_1,x_2)u(x_1)u(x_2)]|_{bb(X)=bb(x)}),(" ",= 4x_1^2 u^2 (x_1) + 4x_2^2 u^2 (x_2) + 8r(x_1,x_2) x_1 x_2 u(x_1) u(x_2).):}
> - u^2(delta y) = {:[({partial f}/{partial ii(X)_1})^2 u^2(x_1) + ({partial f}/{partial ii(X)_2})^2]|_{bb(X)=bb(x)} = 4 x_1^2 u^2 (x_1) + 4 x_2^2 u^2 (x_2),
> - lambda_1 = {|:(b_1 - a_1) - (b_2 - a_2):|}/2, " " " " lambda_2 = {b - a}/2,
> - ii(E)(ii(X)) = {a + b}/2, "    " ii(V)(ii(X)) = {(b - a)^2}/12
> - i_j = (a_j xx i_j) mod d_j
> - i_j = a_j xx (i_j mod b_j) - c_j xx floor(i_j//b_j)
> - [(x_1),(x_2)], " " [(u^2(x_1),ru(x_1)u(x_2)),(ru(x_1)u(x_2),u^2(x_2))].
> - 1 + (ii(rho)_a - ii(rho)_a_0 )(1//ii(rho)_W - 1//ii(rho)_R)
> - 


These are the examples I picked up from the Asciimath specs.

**Comment by suleman-uzair** [comment]
Fixed this issue in #110
